src/mame/drivers/didact.cpp

```diff
@@ -457,34 +451,381 @@ void mp68a_state::machine_start()
+	PORT_BIT(0x04, IP_ACTIVE_LOW,	IPT_KEYBOARD)								PORT_CODE(KEYCODE_Z)  			PORT_CHAR('Z')  PORT_CHAR('z')
+	PORT_BIT(0x08, IP_ACTIVE_LOW,	IPT_KEYBOARD)								PORT_CODE(KEYCODE_A)  			PORT_CHAR('A')  PORT_CHAR('a')
+	PORT_BIT(0x10, IP_ACTIVE_LOW,	IPT_KEYBOARD)								PORT_CODE(KEYCODE_Q)  			PORT_CHAR('q')  PORT_CHAR('Q')
+	PORT_BIT(0x20, IP_ACTIVE_LOW,	IPT_KEYBOARD)								PORT_CODE(KEYCODE_1)  			PORT_CHAR('1')	PORT_CHAR('!')
+	PORT_BIT(0x40, IP_ACTIVE_LOW,	IPT_UNUSED)
+	PORT_BIT(0x80, IP_ACTIVE_LOW,	IPT_UNUSED)
+
+	PORT_START("LINE2")
+	PORT_BIT(0x01, IP_ACTIVE_LOW,	IPT_KEYBOARD)								PORT_CODE(KEYCODE_PLUS_PAD)  	PORT_CHAR('+')
+	PORT_BIT(0x02, IP_ACTIVE_LOW,	IPT_KEYBOARD)								PORT_CODE(KEYCODE_MINUS_PAD)	PORT_CHAR('-')
+	PORT_BIT(0x04, IP_ACTIVE_LOW,	IPT_KEYBOARD)								PORT_CODE(KEYCODE_X)  			PORT_CHAR('x')  PORT_CHAR('X')
+	PORT_BIT(0x08, IP_ACTIVE_LOW,	IPT_KEYBOARD)								PORT_CODE(KEYCODE_S)  			PORT_CHAR('s')  PORT_CHAR('S')
+	PORT_BIT(0x10, IP_ACTIVE_LOW,	IPT_KEYBOARD)								PORT_CODE(KEYCODE_W)  			PORT_CHAR('w')  PORT_CHAR('W')
+	PORT_BIT(0x20, IP_ACTIVE_LOW,	IPT_KEYBOARD)								PORT_CODE(KEYCODE_2)  			PORT_CHAR('2')  PORT_CHAR('"')
+	PORT_BIT(0x40, IP_ACTIVE_LOW,	IPT_KEYBOARD)								PORT_CODE(KEYCODE_SLASH_PAD)  	PORT_CHAR('/')
+	PORT_BIT(0x80, IP_ACTIVE_LOW,	IPT_KEYBOARD)								PORT_CODE(KEYCODE_ASTERISK)		PORT_CHAR('*')
+
+	PORT_START("LINE3")
+	PORT_BIT(0x01, IP_ACTIVE_LOW,	IPT_KEYBOARD)								PORT_CODE(KEYCODE_COLON) 		PORT_CHAR('�') PORT_CHAR('�')
+	PORT_BIT(0x02, IP_ACTIVE_LOW,	IPT_UNUSED)
+	PORT_BIT(0x04, IP_ACTIVE_LOW,	IPT_KEYBOARD)								PORT_CODE(KEYCODE_STOP) 		PORT_CHAR('.') PORT_CHAR(':')
+	PORT_BIT(0x08, IP_ACTIVE_LOW,	IPT_KEYBOARD)								PORT_CODE(KEYCODE_L)  			PORT_CHAR('l')  PORT_CHAR('L')
+	PORT_BIT(0x10, IP_ACTIVE_LOW,	IPT_KEYBOARD)								PORT_CODE(KEYCODE_O)  			PORT_CHAR('o')  PORT_CHAR('O')
+	PORT_BIT(0x20, IP_ACTIVE_LOW,	IPT_KEYBOARD)								PORT_CODE(KEYCODE_9)  			PORT_CHAR(')')  PORT_CHAR('9')
+	PORT_BIT(0x40, IP_ACTIVE_LOW,	IPT_KEYBOARD)								PORT_CODE(KEYCODE_P)  			PORT_CHAR('p')  PORT_CHAR('P')
+	PORT_BIT(0x80, IP_ACTIVE_LOW,	IPT_KEYBOARD)								PORT_CODE(KEYCODE_0)  			PORT_CHAR('0')  PORT_CHAR('=')
+
+	PORT_START("LINE4")
+	PORT_BIT(0x01, IP_ACTIVE_LOW,	IPT_KEYBOARD)								PORT_CODE(KEYCODE_QUOTE)		PORT_CHAR('�') PORT_CHAR('�')
+	PORT_BIT(0x02, IP_ACTIVE_LOW,	IPT_KEYBOARD)								PORT_CODE(KEYCODE_SLASH)		PORT_CHAR('-') PORT_CHAR('_')
+	PORT_BIT(0x04, IP_ACTIVE_LOW,	IPT_KEYBOARD)								PORT_CODE(KEYCODE_COMMA)		PORT_CHAR(',') PORT_CHAR(';')
+	PORT_BIT(0x08, IP_ACTIVE_LOW,	IPT_KEYBOARD)								PORT_CODE(KEYCODE_K)  			PORT_CHAR('k')  PORT_CHAR('K')
+	PORT_BIT(0x10, IP_ACTIVE_LOW,	IPT_KEYBOARD)								PORT_CODE(KEYCODE_I)  			PORT_CHAR('i')  PORT_CHAR('I')
+	PORT_BIT(0x20, IP_ACTIVE_LOW,	IPT_KEYBOARD)								PORT_CODE(KEYCODE_8)  			PORT_CHAR('8')  PORT_CHAR('(')
+	PORT_BIT(0x40, IP_ACTIVE_LOW,	IPT_KEYBOARD)								PORT_CODE(KEYCODE_OPENBRACE) 	PORT_CHAR('�') PORT_CHAR('�')
```

Comment:
Ah, just realised these non-ASCII ones won't work properly with all compilers.  MAME source is UTF-8 but the `PORT_CHAR` macro needs a UCS4 character, so you have to supply it as a hex code (e.g. `PORT_CHAR(0x00c5)` for Å).  You can give the proper character in a comment for clarity, we allow any Unicode in comments.